src/network_engine.cpp

```diff
@@ -191,18 +191,14 @@ NetworkEngine::sendRequest(std::shared_ptr<Request>& request)
 
 /* Rate control for requests we receive. */
 bool
-NetworkEngine::rateLimit()
+NetworkEngine::rateLimit(const SockAddr& addr)
 {
-    using namespace std::chrono;
     const auto& now = scheduler.time();
-    while (not rate_limit_time.empty() and duration_cast<seconds>(now - rate_limit_time.front()) > seconds(1))
-        rate_limit_time.pop();
-
-    if (rate_limit_time.size() >= MAX_REQUESTS_PER_SEC)
+    if (not rate_limiter.limit(now))
         return false;
 
-    rate_limit_time.emplace(now);
-    return true;
+    auto it = address_rate_limiter.emplace(addr, Limiter<MAX_REQUESTS_PER_SEC/8>{});
```

Comment:
emplace will only emplace if the element is not in the map, otherwise will just return the <iterator, bool>. The bool is true if insertion took place.